File: files_cpp/src/rest_client.cpp
#include "rest_client.hpp"
#include "config.hpp"
#include "auth.hpp"
#include "logger.hpp"
#include <curl/curl.h>
#include <nlohmann/json.hpp>
#include <sstream>
#include <iostream>
#include <thread>
#include <cmath>
#include <unordered_set>

using json = nlohmann::json;

namespace depthos {

// Callback for curl to write response data
static size_t write_callback(void* contents, size_t size, size_t nmemb, void* userp) {
    size_t total_size = size * nmemb;
    std::string* response = static_cast<std::string*>(userp);
    response->append(static_cast<char*>(contents), total_size);
    return total_size;
}

RestClient::RestClient() {
    curl_handle_ = curl_easy_init();
    if (!curl_handle_) {
        throw std::runtime_error("Failed to initialize CURL");
    }
}

RestClient::~RestClient() {
    if (curl_handle_) {
        curl_easy_cleanup(curl_handle_);
    }
}

HttpResponse RestClient::request(
    const std::string& method,
    const std::string& path,
    const std::unordered_map<std::string, std::string>& query,
    const std::string& body,
    bool signed_request
) {
    HttpResponse response;
    std::string url = std::string(FUTURES_REST_URL) + path;
    
    // Build query string
    if (!query.empty()) {
        url += "?";
        bool first = true;
        for (const auto& [key, value] : query) {
            if (!first) url += "&";
            url += key + "=" + value;
            first = false;
        }
    }
    
    // Set up CURL
    curl_easy_setopt(curl_handle_, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl_handle_, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl_handle_, CURLOPT_WRITEDATA, &response.body);
    curl_easy_setopt(curl_handle_, CURLOPT_TIMEOUT, 15L);
    
    // Headers
    struct curl_slist* headers = nullptr;
    
    if (signed_request) {
        std::string query_str;
        for (const auto& [key, value] : query) {
            query_str += key + "=" + value + "&";
        }
        if (!query_str.empty()) {
            query_str.pop_back();  // Remove trailing &
        }
        
        auto auth_headers = sign_rest(method, path, query_str, body, API_KEY, API_SECRET);
        for (const auto& [key, value] : auth_headers) {
            std::string header = key + ": " + value;
            headers = curl_slist_append(headers, header.c_str());
        }
    }
    
    headers = curl_slist_append(headers, "Content-Type: application/json");
    headers = curl_slist_append(headers, "Accept: application/json");
    curl_easy_setopt(curl_handle_, CURLOPT_HTTPHEADER, headers);
    
    // Method
    if (method == "POST") {
        curl_easy_setopt(curl_handle_, CURLOPT_POST, 1L);
        curl_easy_setopt(curl_handle_, CURLOPT_POSTFIELDS, body.c_str());
    } else if (method == "DELETE") {
        curl_easy_setopt(curl_handle_, CURLOPT_CUSTOMREQUEST, "DELETE");
    } else {
        curl_easy_setopt(curl_handle_, CURLOPT_HTTPGET, 1L);
    }
    
    // Perform request
    CURLcode res = curl_easy_perform(curl_handle_);
    
    if (res != CURLE_OK) {
        response.error = curl_easy_strerror(res);
        response.status_code = 0;
    } else {
        curl_easy_getinfo(curl_handle_, CURLINFO_RESPONSE_CODE, &response.status_code);
    }
    
    curl_slist_free_all(headers);
    
    // Retry logic for 429/5xx
    const int max_retries = 5;
    const std::vector<int> retry_codes = {429, 500, 502, 503, 504};
    
    for (int attempt = 1; attempt <= max_retries; attempt++) {
        if (response.is_success()) break;
        
        bool should_retry = false;
        for (int code : retry_codes) {
            if (response.status_code == code) {
                should_retry = true;
                break;
            }
        }
        
        if (!should_retry) break;
        
        double delay = 0.5 * std::pow(2, attempt - 1);
        LOG_WARN("HTTP {} on {} {} - retry {} in {}s", response.status_code, method, path, attempt, delay);
        std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(delay * 1000)));
        
        // Retry the request
        response.body.clear();
        res = curl_easy_perform(curl_handle_);
        
        if (res != CURLE_OK) {
            response.error = curl_easy_strerror(res);
            response.status_code = 0;
        } else {
            curl_easy_getinfo(curl_handle_, CURLINFO_RESPONSE_CODE, &response.status_code);
        }
    }
    
    return response;
}

std::unordered_map<std::string, ContractSpec> RestClient::fetch_contract_specs(
    const std::vector<std::string>& contract_names
) {
    std::unordered_map<std::string, ContractSpec> specs;
    
    HttpResponse response = request("GET", "/futures/" + std::string(SETTLE) + "/contracts");
    
    if (!response.is_success()) {
        LOG_ERROR("Failed to fetch contract specs: {}", response.body);
        return specs;
    }
    
    try {
        json data = json::parse(response.body);
        std::unordered_set<std::string> universe(contract_names.begin(), contract_names.end());
        
        for (const auto& c : data) {
            std::string name = c["name"].get<std::string>();
            if (universe.find(name) == universe.end()) continue;
            
            Decimal tick = Decimal(c["order_price_round"].get<std::string>());
            int lot = c["order_size_min"].get<int>();
            Decimal quanto = Decimal(c["quanto_multiplier"].get<std::string>());
            Decimal last = Decimal(c.value<std::string>("last_price", "0"));
            
            // Verify current last price is within micro-price range
            if (last > Decimal("0.10")) {
                LOG_WARN("{} last_price={} exceeds micro-price ceiling, skipping", name, last.str());
                continue;
            }
            
            ContractSpec spec;
            spec.name = name;
            spec.tick_size = tick;
            spec.lot_size = lot;
            spec.quanto_multiplier = quanto;
            spec.max_price = Decimal("0.10");
            
            specs[name] = spec;
            
            LOG_INFO("Loaded spec {}  tick={}  lot={}  quanto={}  last={}", name, tick.str(), lot, quanto.str(), last.str());
        }
    } catch (const json::exception& e) {
        LOG_ERROR("JSON parse error in fetch_contract_specs: {}", e.what());
    }
    
    return specs;
}

std::string RestClient::fetch_account_mode() {
    HttpResponse response = request("GET", "/futures/" + std::string(SETTLE) + "/accounts");
    
    if (!response.is_success()) {
        LOG_ERROR("Failed to fetch account mode");
        return "single";
    }
    
    try {
        json data = json::parse(response.body);
        std::string mode = data.value<std::string>("mode", "single");
        LOG_INFO("Account mode: {}", mode);
        return mode;
    } catch (const json::exception& e) {
        LOG_ERROR("JSON parse error in fetch_account_mode: {}", e.what());
        return "single";
    }
}

Decimal RestClient::fetch_balance() {
    HttpResponse response = request("GET", "/futures/" + std::string(SETTLE) + "/accounts");
    
    if (!response.is_success()) {
        return Decimal("0");
    }
    
    try {
        json data = json::parse(response.body);
        std::string available = data.value<std::string>("available", "0");
        Decimal balance = Decimal(available);
        LOG_INFO("USDT Futures balance: {}", balance.str());
        return balance;
    } catch (const json::exception& e) {
        LOG_ERROR("JSON parse error in fetch_balance: {}", e.what());
        return Decimal("0");
    }
}

int RestClient::fetch_position(const std::string& contract) {
    std::string path;
    if (mm_config.account_mode == "dual") {
        path = "/futures/" + std::string(SETTLE) + "/dual_positions/" + contract;
    } else {
        path = "/futures/" + std::string(SETTLE) + "/positions/" + contract;
    }
    
    HttpResponse response = request("GET", path);
    
    if (!response.is_success()) {
        return 0;
    }
    
    try {
        json data = json::parse(response.body);
        
        if (mm_config.account_mode == "dual") {
            int long_size = data.value<json>("long", json::object()).value<int>("size", 0);
            int short_size = data.value<json>("short", json::object()).value<int>("size", 0);
            return long_size - short_size;
        } else {
            return data.value<int>("size", 0);
        }
    } catch (const json::exception& e) {
        LOG_ERROR("JSON parse error in fetch_position: {}", e.what());
        return 0;
    }
}

std::string RestClient::place_order(
    const std::string& contract,
    int size,
    const Decimal& price,
    const std::string& tif,
    bool reduce_only,
    const std::string& text
) {
    if (mm_config.dry_run) {
        LOG_DEBUG("[DRY-RUN] place_order contract={} size={} price={} tif={}", contract, size, price.str(), tif);
        return "-1";
    }
    
    std::string path = "/futures/" + std::string(SETTLE) + "/orders";
    ContractSpec spec = mm_config.contracts[contract];
    
    // Round price to tick size
    Decimal rounded_price = round_to_tick(price, spec.tick_size);
    
    // Build JSON body
    json body_json;
    body_json["contract"] = contract;
    body_json["size"] = size;
    body_json["price"] = rounded_price.str();
    body_json["tif"] = tif;
    body_json["text"] = text;
    body_json["reduce_only"] = reduce_only;
    
    if (mm_config.account_mode == "dual" && !reduce_only) {
        body_json["auto_size"] = (size < 0) ? "close_long" : "close_short";
    }
    
    std::string body = body_json.dump();
    
    HttpResponse response = request("POST", path, {}, body);
    
    if (!response.is_success()) {
        LOG_ERROR("Failed to place order: {}", response.body);
        return "";
    }
    
    try {
        json data = json::parse(response.body);
        std::string id = data.value<std::string>("id", "");
        LOG_INFO("Placed order id={} contract={} size={} price={} status={}", id, contract, size, rounded_price.str(), data.value<std::string>("status", ""));
        return id;
    } catch (const json::exception& e) {
        LOG_ERROR("JSON parse error in place_order: {}", e.what());
        return "";
    }
}

bool RestClient::cancel_order(int order_id, const std::string& contract) {
    if (mm_config.dry_run) {
        LOG_DEBUG("[DRY-RUN] cancel_order id={}", order_id);
        return true;
    }
    
    std::string path = "/futures/" + std::string(SETTLE) + "/orders/" + std::to_string(order_id);
    HttpResponse response = request("DELETE", path);
    
    if (!response.is_success()) {
        LOG_ERROR("Failed to cancel order: {}", response.body);
        return false;
    }
    
    try {
        json data = json::parse(response.body);
        std::string id = data.value<std::string>("id", "");
        LOG_INFO("Canceled order id={} contract={} status={}", id, contract, data.value<std::string>("status", ""));
        return true;
    } catch (const json::exception& e) {
        LOG_ERROR("JSON parse error in cancel_order: {}", e.what());
        return false;
    }
}

int RestClient::cancel_all_orders(const std::string& contract) {
    if (mm_config.dry_run) {
        LOG_DEBUG("[DRY-RUN] cancel_all {}", contract);
        return 0;
    }
    
    std::string path = "/futures/" + std::string(SETTLE) + "/orders";
    std::unordered_map<std::string, std::string> query = {
        {"contract", contract},
        {"status", "open"}
    };
    
    HttpResponse response = request("DELETE", path, query);
    
    if (!response.is_success()) {
        return 0;
    }
    
    // Parse JSON to get count
    return 0;  // Placeholder
}

std::vector<std::unordered_map<std::string, std::string>> RestClient::list_open_orders(
    const std::string& contract
) {
    std::string path = "/futures/" + std::string(SETTLE) + "/orders";
    std::unordered_map<std::string, std::string> query = {
        {"contract", contract},
        {"status", "open"},
        {"limit", "100"}
    };
    
    HttpResponse response = request("GET", path, query);
    
    // Parse JSON and return list
    return {};  // Placeholder
}

Decimal RestClient::round_to_tick(const Decimal& price, const Decimal& tick) {
    // Floor price to nearest tick size
    return floor(price / tick) * tick;
}

} // namespace depthos

File: files/rest_client.py
"""
rest_client.py — Gate.io Futures REST client.

Responsibilities:
  • Bootstrap: fetch contract specs (tick size, lot size, quanto multiplier).
  • Order lifecycle: create, cancel, query individual and bulk.
  • Account: query position, balance, account mode.

All methods are async and return typed dataclasses or raw dicts.
Retries with exponential back-off on 429 / 5xx.
"""
from __future__ import annotations

import asyncio
import json
import logging
import time
from decimal import Decimal
from typing import Any, Dict, List, Optional, Tuple

import aiohttp

from auth import sign_rest, API_KEY, API_SECRET
from config import FUTURES_REST_URL, SETTLE, ContractSpec, mm_config

log = logging.getLogger("rest_client")

# ─── Retry policy ───────────────────────────────────────────────────────────
_MAX_RETRIES   = 5
_BASE_DELAY_S  = 0.5
_RETRY_CODES   = {429, 500, 502, 503, 504}


async def _request(
    session: aiohttp.ClientSession,
    method: str,
    path: str,
    *,
    query: Optional[Dict[str, Any]] = None,
    body: Optional[Dict[str, Any]]  = None,
    signed: bool = True,
) -> Any:
    """
    Core HTTP request with retry logic.

    Returns parsed JSON body on 2xx, raises on non-retryable errors.
    """
    url = FUTURES_REST_URL + path
    headers: Dict[str, str] = {}

    if signed:
        headers = sign_rest(
            method.upper(), path, query, body,
            API_KEY, API_SECRET,
        )

    body_bytes = json.dumps(body, separators=(",", ":")).encode() if body else None

    for attempt in range(1, _MAX_RETRIES + 1):
        try:
            async with session.request(
                method,
                url,
                headers=headers,
                params=query,
                data=body_bytes,
            ) as resp:
                raw = await resp.json(content_type=None)
                if resp.status == 200:
                    return raw
                if resp.status in _RETRY_CODES and attempt < _MAX_RETRIES:
                    delay = _BASE_DELAY_S * (2 ** (attempt - 1))
                    log.warning(
                        "HTTP %s on %s %s — retry %d in %.2fs",
                        resp.status, method, path, attempt, delay,
                    )
                    await asyncio.sleep(delay)
                    continue
                log.error(
                    "REST error HTTP %s: %s", resp.status, raw
                )
                raise RuntimeError(f"REST {method} {path} → HTTP {resp.status}: {raw}")
        except aiohttp.ClientError as exc:
            if attempt < _MAX_RETRIES:
                await asyncio.sleep(_BASE_DELAY_S * (2 ** attempt))
            else:
                raise

    raise RuntimeError(f"REST {method} {path} exhausted {_MAX_RETRIES} retries")


# ─── Bootstrap ───────────────────────────────────────────────────────────────

async def fetch_contract_specs(
    session: aiohttp.ClientSession,
    contract_names: List[str],
) -> Dict[str, ContractSpec]:
    """
    Fetch and validate contract specs from Gate.io.
    Returns only contracts whose mark price is currently ≤ $0.10.
    """
    path = f"/futures/{SETTLE}/contracts"
    all_contracts: List[Dict] = await _request(session, "GET", path, signed=False)

    specs: Dict[str, ContractSpec] = {}
    universe = set(contract_names)

    for c in all_contracts:
        name = c["name"]
        if name not in universe:
            continue

        tick  = Decimal(str(c["order_price_round"]))
        lot   = int(c["order_size_min"])
        quanto = Decimal(str(c["quanto_multiplier"]))

        # Verify current last price is within micro-price range.
        last = Decimal(str(c.get("last_price", "0")))
        if last > Decimal("0.10"):
            log.warning(
                "%s last_price=%s exceeds micro-price ceiling, skipping",
                name, last,
            )
            continue

        specs[name] = ContractSpec(
            name=name,
            tick_size=tick,
            lot_size=lot,
            quanto_multiplier=quanto,
        )
        log.info(
            "Loaded spec %-20s  tick=%s  lot=%s  quanto=%s  last=%s",
            name, tick, lot, quanto, last,
        )

    return specs


async def fetch_account_mode(session: aiohttp.ClientSession) -> str:
    """
    Returns 'single' (one-way) or 'dual' (hedge mode).
    """
    path = f"/futures/{SETTLE}/accounts"
    data = await _request(session, "GET", path)
    mode = data.get("mode", "single")
    log.info("Account mode: %s", mode)
    return mode


async def fetch_balance(session: aiohttp.ClientSession) -> Decimal:
    """Returns total USDT available balance in futures account."""
    path = f"/futures/{SETTLE}/accounts"
    data = await _request(session, "GET", path)
    return Decimal(str(data.get("available", "0")))


async def fetch_position(
    session: aiohttp.ClientSession,
    contract: str,
) -> int:
    """
    Returns current net position in contracts (positive=long, negative=short).
    Handles both single and dual account modes.
    """
    if mm_config.account_mode == "dual":
        path = f"/futures/{SETTLE}/dual_positions/{contract}"
        data = await _request(session, "GET", path)
        long_size  = int(data.get("long",  {}).get("size", 0))
        short_size = int(data.get("short", {}).get("size", 0))
        return long_size - short_size
    else:
        path = f"/futures/{SETTLE}/positions/{contract}"
        data = await _request(session, "GET", path)
        return int(data.get("size", 0))


# ─── Order management ────────────────────────────────────────────────────────

async def place_order(
    session:   aiohttp.ClientSession,
    contract:  str,
    size:      int,          # positive=buy, negative=sell
    price:     Decimal,
    tif:       str = "gtc",  # gtc | ioc | poc (post-only)
    reduce_only: bool = False,
    text:      str  = "t-mm",
) -> Dict[str, Any]:
    """
    Place a single limit order on Gate.io futures.

    Gate.io conventions:
      • size > 0 → buy (long open / short close)
      • size < 0 → sell (short open / long close)
      • price '0' = market (not used here)
      • tif 'poc' = post-only (maker-or-cancel)

    Returns the raw order dict from the exchange.
    """
    if mm_config.dry_run:
        log.info(
            "[DRY-RUN] place_order contract=%s size=%+d price=%s tif=%s",
            contract, size, price, tif,
        )
        return {"id": -1, "status": "open", "dry_run": True}

    path = f"/futures/{SETTLE}/orders"
    spec: ContractSpec = mm_config.contracts[contract]

    # Round price to tick size.
    price = _round_to_tick(price, spec.tick_size)

    payload: Dict[str, Any] = {
        "contract":    contract,
        "size":        size,
        "price":       str(price),
        "tif":         tif,
        "text":        text,
        "reduce_only": reduce_only,
    }

    if mm_config.account_mode == "dual":
        # Dual mode: Gate.io requires explicit open/close and side.
        if not reduce_only:
            payload["auto_size"] = "close_long" if size < 0 else "close_short"

    resp = await _request(session, "POST", path, body=payload)
    log.debug(
        "Placed order id=%s contract=%s size=%+d price=%s status=%s",
        resp.get("id"), contract, size, price, resp.get("status"),
    )
    return resp


async def cancel_order(
    session:    aiohttp.ClientSession,
    order_id:   int,
    contract:   str,
) -> bool:
    """
    Cancel a single order. Returns True if successfully cancelled.
    Swallows 404 (already filled/cancelled) and returns False.
    """
    if mm_config.dry_run:
        log.info("[DRY-RUN] cancel_order id=%s", order_id)
        return True

    path = f"/futures/{SETTLE}/orders/{order_id}"
    try:
        resp = await _request(session, "DELETE", path)
        status = resp.get("status", "")
        log.debug("Cancelled order id=%s contract=%s status=%s", order_id, contract, status)
        return status in ("cancelled", "closed")
    except RuntimeError as exc:
        if "404" in str(exc):
            log.debug("cancel_order: order %s already gone", order_id)
            return False
        raise


async def cancel_all_orders(
    session:   aiohttp.ClientSession,
    contract:  str,
) -> int:
    """
    Cancel all open orders for a contract.
    Returns count of cancelled orders.
    """
    if mm_config.dry_run:
        log.info("[DRY-RUN] cancel_all_orders contract=%s", contract)
        return 0

    path  = f"/futures/{SETTLE}/orders"
    query = {"contract": contract, "status": "open"}
    resp  = await _request(session, "DELETE", path, query=query)
    n = len(resp) if isinstance(resp, list) else 0
    log.info("Cancelled %d orders for %s", n, contract)
    return n


async def list_open_orders(
    session:  aiohttp.ClientSession,
    contract: str,
) -> List[Dict[str, Any]]:
    """List all open orders for a contract."""
    path  = f"/futures/{SETTLE}/orders"
    query = {"contract": contract, "status": "open", "limit": 100}
    return await _request(session, "GET", path, query=query)


# ─── Helpers ─────────────────────────────────────────────────────────────────

def _round_to_tick(price: Decimal, tick: Decimal) -> Decimal:
    """
    Floor price to nearest tick size.
    For a sell we floor; for a buy we also floor — joining the level.
    """
    return (price // tick) * tick
